spec/controllers/single_page_subscriptions_controller_spec.rb

```diff
@@ -0,0 +1,122 @@
+        expect(response).to redirect_to("#{auth_provider}?redirect_path=#{redirect_path}&_ga=abc123&cookie_consent=accept")
+      end
+    end
+
+    describe "POST /email/subscriptions/single-page/new" do
+      before { stub_content_store_has_item(topic) }
+
+      it "404s when a content id cannot be found" do
+        stub_content_store_does_not_have_item(topic)
+        get :show, params: params
+        expect(response).to have_http_status(:not_found)
+      end
+
+      context "when a user is not logged in" do
+        it "renders the view with a sign in link including the topic" do
+          get :show, params: params
+          expect(response.body).to include(single_page_new_session_path)
+        end
+      end
+
+      context "when a user is logged in" do
+        let(:session_id) { "session-id" }
+        let(:user_id) { "user-id" }
+        let(:subscriber_id) { "subscriber-id" }
+        let(:user_email) { "test@gov.uk" }
+
+        before do
+          mock_logged_in_session(session_id)
+          stub_account_api_user_info(id: user_id)
+          stub_account_api_get_sign_in_url(auth_uri: auth_provider, redirect_path: redirect_path)
+          stub_email_alert_api_creates_subscriber_list({
+            url: topic,
+            title: topic_name,
+            slug: topic_slug,
+          })
+
+          stub_account_api_put_email_subscription(
+            name: topic_name,
+            topic_slug: topic_slug,
+          )
+        end
+
+        it "logs the user out if the session is invalid" do
+          stub_account_api_unauthorized_user_info
+          post :show, params: params
+          expect(response).to redirect_to("#{auth_provider}?redirect_path=#{redirect_path}")
+        end
+
+        context "and they have no unlinked subscriptions" do
+          let(:subscriber_list_id) { SecureRandom.uuid }
+
+          it "subscribes them and redirects back to the page" do
+            stub_email_alert_api_find_subscriber_by_govuk_account(
+              user_id, subscriber_id, user_email
+            )
+            post :show, params: params
+            expect(response).to redirect_to("http://test.host#{topic}")
+          end
+        end
+
+        context "and they have unlinked subscriptions" do
+          let(:subscriber_id) { "subscriber-id" }
+          it "merges their accounts, subscribes them to the topic and redirects back to the page" do
+            stub_email_alert_api_find_subscriber_by_govuk_account_no_subscriber(user_id)
+            stub_email_alert_api_link_subscriber_to_govuk_account(
+              session_id,
+              subscriber_id,
+              user_email,
+              govuk_account_id: user_id,
+            )
+            post :show, params: params
+            expect(response).to redirect_to("http://test.host#{topic}")
+          end
+        end
+      end
+    end
+  end
+end
```

Comment:
Nitpick: if you have a `context` it should have some `let`s or a `before` so it's clear what precondition it's describing.  Don't just put the precondition in the test body: it's easier to miss, and also won't work if you add a second test.

Reply:
@alex9smith and I had a chat about this, what do you think Alex?
We talked a bit about doing it this way round vs in before blocks?

Reply:
Yeah we went back and forth on this this for the last two tests. We could pull out the `stub_*` calls in each test to be a `before`:

```
context "and they have unlinked subscriptions" do
  before do
    stub_email_alert_api_find_subscriber_by_govuk_account_no_subscriber(user_id)
    stub_email_alert_api_link_subscriber_to_govuk_account(
      session_id,
      subscriber_id,
      user_email,
      govuk_account_id: user_id,
    )
  end

  it "merges their accounts, subscribes them to the topic and redirects back to the page" do
    post :show, params: params
    expect(response).to redirect_to("http://test.host#{topic}")
end
```

But then that leaves the contents of the `it` blocks the same for each test which we thought was a bit weird and confusing given they are testing different paths in the logic.